lib/encrypt/encrypt.go

```diff
@@ -73,40 +76,104 @@ func NewCipher(method, password string) (*Cipher, error) {
 	if password == "" {
 		return nil, errors.New("password required")
 	}
-	info, ok := GetCipherInfo(method)
-	if !ok {
-		return nil, errors.New("Unsupported encryption method: " + method)
-	}
-	key := Evp2Key(password, info.KeyLen)
 
-	// hash(key) -> read IV
-	riv := sha256.New().Sum(key)[:info.IvLen]
-	rd, err := info.NewStream(key, riv, false)
-	if err != nil {
-		return nil, err
+	if info, ok := complexCiphers[method]; ok {
+		key := Evp2Key(password, info.keyLen)
+
+		// hash(key) -> read IV
+		riv := sha256.New().Sum(key)[:info.IvLen()]
+		rd, err := info.newStream(&encDec{key, riv, info.newCipher, info.newEncrypt})
+		if err != nil {
+			return nil, err
+		}
+		// hash(read IV) -> write IV
+		wiv := sha256.New().Sum(riv)[:info.IvLen()]
+		wr, err := info.newStream(&encDec{key, wiv, info.newCipher, info.newDecrypt})
+		if err != nil {
+			return nil, err
+		}
+		return &Cipher{wr, rd}, nil
 	}
-	// hash(read IV) -> write IV
-	wiv := sha256.New().Sum(riv)[:info.IvLen]
-	wr, err := info.NewStream(key, wiv, true)
-	if err != nil {
-		return nil, err
+
+	if info, ok := simpleCiphers[method]; ok {
+		key := Evp2Key(password, info.keyLen)
+
+		// hash(key) -> read IV
+		riv := sha256.New().Sum(key)[:info.IvLen()]
+		wr, err := info.newStream(key[:info.keyLen], riv[:info.ivLen])
+		if err != nil {
+			return nil, err
+		}
+		// hash(read IV) -> write IV
+		wiv := sha256.New().Sum(riv)[:info.IvLen()]
+		rd, err := info.newStream(key[:info.keyLen], wiv[:info.ivLen])
+		if err != nil {
+			return nil, err
+		}
+		return &Cipher{wr, rd}, nil
 	}
-	return &Cipher{wr, rd}, nil
+	return nil, errors.New("unsupported encryption method: " + method)
 }
 
 // NewStream new stream
 func NewStream(method string, key, iv []byte, encrypt bool) (cipher.Stream, error) {
-	info, ok := GetCipherInfo(method)
-	if !ok {
-		return nil, errors.New("Unsupported encryption method: " + method)
+	check := func(info KeyIvLen) error {
+		if len(key) < info.KeyLen() {
+			return errors.New("invalid key size " + strconv.Itoa(len(key)))
+		}
+		if len(iv) < info.IvLen() {
+			return errors.New("invalid IV length " + strconv.Itoa(len(iv)))
+		}
+		return nil
+	}
+
+	if info, ok := complexCiphers[method]; ok {
+		if err := check(info); err != nil {
+			return nil, err
+		}
+		encdec := info.newDecrypt
+		if encrypt {
+			encdec = info.newEncrypt
+		}
+		return info.newStream(&encDec{key[:info.keyLen], iv[:info.ivLen], info.newCipher, encdec})
+	}
+
+	if info, ok := simpleCiphers[method]; ok {
+		if err := check(info); err != nil {
+			return nil, err
+		}
+		return info.newStream(key[:info.keyLen], iv[:info.ivLen])
 	}
-	if len(key) < info.KeyLen {
-		return nil, errors.New("invalid key size " + strconv.Itoa(len(key)))
+	return nil, errors.New("unsupported encryption method: " + method)
+}
+
+// GetCipherInfo 根据方法获得 Cipher information
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
comment on exported function GetCipher should be of the form "GetCipher ..."